networking_ccloud/ml2/agent/eos/switch.py

```diff
@@ -32,6 +31,116 @@
+        IFACE_PC_VTRANS_EGRESS = ("interfaces/interface[name={iface}]/aggregation/switched-vlan/vlan-translation/"
+                                  "egress[translation-key={vlan}]")
+        IFACE_PC_VTRANS_INGRESS = ("interfaces/interface[name={iface}]/aggregation/switched-vlan/vlan-translation/"
+                                   "ingress[translation-key={vlan}]")
+
+    def __init__(self, switch_name, *args, **kwargs):
+        self._gnmi = gNMIclient(*args, **kwargs)
+        self._switch_name = switch_name
+
+    def connect(self, *args, **kwargs):
+        try:
+            self._gnmi.close()  # doesn't raise on already closed connection
+        except AttributeError:
+            # probably never connected (no channel present)
+            pass
+        # FIXME: currently this throws Except (not retriable)
+        #   would possibly raise a ConnectionRefusedError
+        #   ConnectionRefusedError: [Errno 111] Connection refused
+        LOG.debug("Connecting to switch %s", self._switch_name)
+        try:
+            self._gnmi.connect(*args, **kwargs)
+        except Exception as e:
+            raise cc_exc.SwitchConnectionError(f"{self._switch_name} connect() {e.__class__.__name__} {e}")
+
+    def _run_method(self, method, *args, retries=3, **kwargs):
+        try:
+            start_time = time.time()
+            data = getattr(self._gnmi, method)(*args, **kwargs)
+            LOG.debug("Command %s() succeeded in %.2fs", method, time.time() - start_time)
+            return data
+        except Exception as e:
+            LOG.exception("Command %s() failed on %s in %.2fs (retries left: %s): %s %s",
+                          method, self._switch_name, time.time() - start_time, retries, e.__class__.__name__, str(e))
+
+            cmd = [f"{method}("]
+            if args:
+                cmd.append(", ".join(map(json.dumps, args)))
+            if kwargs:
+                if args:
+                    cmd.append(", ")
+                cmd.append(", ".join(f"{k}={json.dumps(v)}" for k, v in kwargs.items()))
+            cmd.append(")")
+            cmd = "".join(cmd)
+            LOG.debug("Failed command on %s was %s", self._switch_name, cmd)
+
+            if isinstance(e, AttributeError):
+                # module 'grpc' has no attribute '_channel', sometimes a first-connect problem
+                LOG.info("Reconnecting %s because of %s %s", self._switch_name, e.__class__.__name__, str(e))
+                self.connect()
+            if retries > 0:
+                time.sleep(0.5)
+                return self._run_method(method, *args, retries=retries - 1, **kwargs)
+            raise cc_exc.SwitchConnectionError(f"{self._switch_name} {method}() {e.__class__.__name__} {e}")
+
+    def get(self, *args, unpack=True, single=True, **kwargs):
+        data = self._run_method("get", *args, **kwargs)
+        if data and unpack:
+            data = data['notification']
+            if single:
```

Comment:
can we also unpack further if `single=False` like:
```
            if single:
                data = data[0]['update'][0]['val']
            else:
                data = [x['update'][0]['val'] for x in data]
```

Reply:
Not sure if we can, we'd have to try it with sysdb. With sysdb things are.... somehwat weird, the structure is a bit different and it doesn't play well with single. We could try though.